### DISCUSSION TITLE
Replace single Node with two new nodes

### DISCUSSION THREAD

**Discussion by mglaman** [rectorphp/rector#6538] [Q&A]
We have a deprecated method that is actually replaced with two different method calls. I'm not sure how to return as returning an array from MethodCall_ node throws an error in the traverse:

Rule: https://github.com/palantirnet/drupal-rector/blob/101d36381b953a44eb2c34c6f1f9e6d1530d3209/src/Rector/Deprecation/UiHelperTraitDrupalPostFormRector.php

Code sample example:

```
            new CodeSample(
                <<<'CODE_BEFORE'
$edit = [];
$edit['action'] = 'action_goto_action';
$this->drupalPostForm('admin/config/system/actions', $edit, 'Create');
CODE_BEFORE
                ,
                <<<'CODE_AFTER'
$edit = [];
$edit['action'] = 'action_goto_action';
$this->drupalGet('admin/config/system/actions');
$this->submitForm($edit, 'Create');
CODE_AFTER
            )
        ]);
```

With this:

```
    public function getNodeTypes(): array
    {
        return [
            Node\Expr\MethodCall::class,
        ];
    }
```

It errors if I return an array and not a single node. Maybe there is a node class I am missing?

**Comment by TomasVotruba** [comment]
Hi, could you provide full rule?

**Reply by mglaman** [reply]
Yes, sorry!

https://github.com/palantirnet/drupal-rector/blob/101d36381b953a44eb2c34c6f1f9e6d1530d3209/src/Rector/Deprecation/UiHelperTraitDrupalPostFormRector.php

**Reply by TomasVotruba** [reply]
I see, thanks!

**Comment by mglaman** [comment]
I just found https://github.com/rectorphp/rector-laravel/blob/main/src/Rector/MethodCall/ChangeQueryWhereDateValueWithCarbonRector.php!

I need to call `addNodeBeforeNode` to add a node before the replaced node. Although, it is deprecated, and `$nodesToAddCollector` is private.

This solved my problem:

```
            if ($options === null) {
                $this->addNodeBeforeNode(
                    $this->nodeFactory->createLocalMethodCall('drupalGet', [$path]),
                    $node
                );
            } else {
                $this->addNodeBeforeNode(
                    $this->nodeFactory->createLocalMethodCall('drupalGet', [$path, $options]),
                    $node
                );
            }
```

**Reply by TomasVotruba** [reply]
That is old approach before we fixed one node by many replace pattern in Rector. It will be slowly removed from Rector as it can cause undesired side effects.

It's more clean to return nodes directly, e.g. like here
https://github.com/rectorphp/rector-src/blob/c277e7d35c3dd930660307c23ec3c74fcad337ab/rules/Defluent/Rector/Return_/DefluentReturnMethodCallRector.php#L78

**Reply by mglaman** [reply]
I'll try that again. When I did, I got an exception from the node traverser and returning an array of nodes when the source wasn't an array. Pulling from memory, but I will try it again.

**Reply by TomasVotruba** [reply]
It might be case if you're trying to replace `MethodCall` with 2 `MethodCall`s. Then node traversers further and replaces a new `MethodCall` with 2 new `MethodCall`s`. And then again and again.

In your case it might help to check the method name.

**Reply by mglaman** [reply]
We get this error when I try to return an array. We have to replace one method call with two.

```
LogicException : leaveNode() may only return an array if the parent structure is an array
```

We have:

```
    public function refactor(Node $node): ?array
    {
        assert($node instanceof Node\Expr\MethodCall);
        if ($this->getName($node->name) === 'drupalPostForm') {
```

Which I return

```
            return [$drupalGetNode, $submitFormNode];
```

With `NodesToAddCollector` private, should I just inject it to our rule?

**Reply by TomasVotruba** [reply]
That's a different case then. Could you share the rule in a PR I can check out?

**Reply by mglaman** [reply]
@TomasVotruba the PR is here: https://github.com/palantirnet/drupal-rector/pull/162. There are _a lot_ of changes. Hopefully this links you right to the file: https://github.com/palantirnet/drupal-rector/pull/162/files#diff-9b4f9d938e5e7eefe227001a5d75bd25d2147a2cb4584d0c401ce82b0d988c05

It's replacing this deprecated method: https://git.drupalcode.org/project/drupal/-/blob/9.3.x/core/tests/Drupal/Tests/UiHelperTrait.php#L198

We need to replace calls to `drupalPostForm` to a `drupalGet` and `submitForm` – the helper method which invoked these two commands methods was deprecated for a more explicit execution. That's why I need to insert a node before the main call.

**Reply by mglaman** [reply]
@TomasVotruba just came back on this a year later to try and fix usages of deprecated `addNodeBeforeNode` and hitting the same problem for the same rule. It looks like rector-laravel is still using it as well: https://github.com/rectorphp/rector-laravel/search?q=addNodeBeforeNode. Will this be safe to use for a while, or will the laravel package have a good example for the fix?

**Reply by TomasVotruba** [reply]
In case both nodes are stmts, they should be returned instead. If one of them is expr, there is no other way at the moment than `addNodeBeforeNode()`.

**Reply by mglaman** [reply]
Thank you!
